### DISCUSSION TITLE
Cache Remember condition

### DISCUSSION THREAD

**Discussion by acip** [laravel/framework#48890] [Ideas]
I recently embarked on a project refactoring task and stumbled upon a scenario where caching null values for a particular `remember` call wasn't beneficial. This sparked an idea to enhance the remember functionality in Laravel. I propose the following extensions:

1. Enable `remember` only if the callback returns a non-null value.
2. Allow `remember` to be conditioned on custom logic.

These enhancements could be implemented either through a new method (for the first suggestion) or by introducing a new parameter to the existing `remember` method. What are your thoughts on this?

**Comment by henzeb** [comment]
I'd suggest a remberIf and a rememberUnless method.

**Comment by gkmk** [comment]
Null values shouldn't be cached by default in Laravel. What version are you using? The [remember](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Cache/Repository.php#L386) function would execute the callback everytime if there is no key or value is null.

**Comment by kfriars** [comment]
I have come across a similar scenario, where remembering `null` values triggers expensive computations on every request.

For redis for example, we have the `exists` method, which tells you if the cache key exists. In my opinion, It would make more sense to return whatever value is stored in cache when the cache key exists, rather than based on the `truthiness` of the value stored.

Adding a new cache method like `persist` or `rememberKey` which does the same thing as remember but depends only on cache key presence would be useful in any case where your remembered closure can rightfully have a null return value.

**Reply by henzeb** [reply]
you could return false. then it gets cached, but with false. adding `?:null`  when returning the response of remember solves your issue.

**Reply by kfriars** [reply]
I hear you @henzeb, but that is more of a workaround to a use-case I imagine is fairly common and people just aren't aware.

Just trying to say there is a good case for an alternate `remember` in a language/framework where null values are quite common.

**Reply by henzeb** [reply]
I do agree. It's just how laravel has setup the Cache api. The "has" method for example just tries to retrieve the he value and tests if it is not null. So this new method would be a workaround on its own...

**Comment by shaedrich** [comment]
So, if _any_ value has to be able to be returned, it doesn't matter, whether we change `null` to `false` or keep `null`. To solve this, there are other options:
* Return `Optional`
* Treat throwing an exception as a non-cacheable return value
